### PR TITLE
DMP-3634: ARM RPO - Replay Failed ARM Objects

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
+*Steps:*+
 * Flyway script for the job is already done in DMP-3807.
 * Add the configuration placeholder in application.yaml and create the corresponding Configuration file for the automated job.
 * Create ARMRpoReplayAutomatedTask in Java. 
 * Fetch rpo_csv_start_hour & rpo_csv_end_hour from arm_automated_task table for *ProcessE2EARMRPOPendingAutomatedTasks* job.
 * The job will look for ARM record in EOD table with status as ARM_RPO_REPLAY and last_updated_ts is within arm_replay_start_ts and arm_replay_end_ts of automated_task table.
 * Reset the status for those records to ARM_RAW_DATA_FAILED (14) and transfer attempt to 0

+*Acceptance Criteria:*+

*AC1: Successful flow - EOD has records with ARM_RPO_REPLAY* 

GIVEN

There exists EOD records with ARM_RPO_REPLAY in external_object_directory table

WHEN

the ARMRpoReplayAutomatedTask runs and completes successfully

THEN

in the EOD table the status of those record will be updated to ARM_RAW_DATA_FAILED(14) and transfer_attempt to 0.

 

*AC2:* *EOD has records with ARM_RPO_REPLAY* **

GIVEN

There exists no EOD records with ARM_RPO_REPLAY in external_object_directory table

WHEN

the ARMRpoReplayAutomatedTask runs and completes successfully

THEN

no records in EOD gets updated.

 

 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by hemantasharma1129 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/common/repository/ExternalObjectDirectoryRepository.java [hunk 1]
```diff
@@ -539,6 +539,20 @@ void updateByStatusEqualsAndDataIngestionTsBefore(ObjectRecordStatusEntity curre
...
+
+    @Query("""
+            update ExternalObjectDirectoryEntity eod
+            set eod.status = :newStatus,
+            eod.transferAttempts = :transferAttempts
+            where eod.status = :oldStatus
+            and eod.lastModifiedDateTime between :startTime and :endTime
+        """)
+    @Modifying
+    void updateEodStatusAndTransferAttempsWhereLastModifiedIsBetweenTwoDateTimesAndHasStatus(
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/repository/ExternalObjectDirectoryRepository.java:551]: Updated

File: src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java [hunk 2]
```diff
@@ -205,6 +206,13 @@ public AutomatedTaskStatus getAutomatedTaskStatus(String taskName) {
         throw new DartsApiException(FAILED_TO_FIND_AUTOMATED_TASK);
     }
 
+    @Override
+    public ArmAutomatedTaskEntity getArmAutomatedTaskEntity(AutomatedTaskName automatedTaskName) {
+        return automatedTaskRepository.findByTaskName(automatedTaskName.getTaskName())
+            .map(automatedTaskEntity -> automatedTaskEntity.getArmAutomatedTaskEntity())
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java:212]: Updated

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -515,6 +515,11 @@ darts:
         lock:
           at-least-for: PT1M
           at-most-for: PT90M
+      arm-rpo-replay:
+        system-user-email: system_ArmRpoReplay@hmcts.net
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:521]: This is intentional :) and is done in preparation for: https://tools.hmcts.net/jira/browse/DMP-4219 in which every automated task will be given a unique user

File: src/test/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImplTest.java [hunk 4]
```diff
@@ -479,6 +481,55 @@ public void run() {
...
+    void positiveGetArmAutomatedTaskEntity() {
+        AutomatedTaskEntity automatedTask = mock(AutomatedTaskEntity.class);
+        ArmAutomatedTaskEntity armAutomatedTaskEntity = mock(ArmAutomatedTaskEntity.class);
+        when(automatedTask.getArmAutomatedTaskEntity()).thenReturn(armAutomatedTaskEntity);
+        when(mockAutomatedTaskRepository.findByTaskName("Test")).thenReturn(Optional.of(automatedTask));
+
+        AutomatedTaskName automatedTaskName = mock(AutomatedTaskName.class);
+        when(automatedTaskName.getTaskName()).thenReturn("Test");
+
+        assertEquals(armAutomatedTaskEntity, automatedTaskService.getArmAutomatedTaskEntity(automatedTaskName));
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImplTest.java:495]: I have split out the verification from the execution

**Review by chrisbellingham-hmcts — APPROVED** [review]

**Pull request merged**